postgres-protocol/src/message/backend.rs

```diff
@@ -340,7 +341,12 @@ impl ReplicationMessage<Bytes> {
             XLOG_DATA_TAG => {
                 let wal_start = buf.read_u64::<BigEndian>()?;
                 let wal_end = buf.read_u64::<BigEndian>()?;
-                let timestamp = buf.read_i64::<BigEndian>()?;
+                let ts = buf.read_i64::<BigEndian>()?;
+                let timestamp = if ts > 0 {
+                    *PG_EPOCH + Duration::from_micros(ts as u64)
+                } else {
+                    *PG_EPOCH - Duration::from_micros(-ts as u64)
+                };
```

Comment:
Going backward from `PG_EPOCH`: is that really allowed?

Perhaps an attempt to represent time before `PG_EPOCH` should just be a hard error?

Reply:
Yeah it's ok to send negative values. This is the pg debug log (on 13.2) when sending UNIX_EPOCH:

```
postgres_1  | 2021-04-21 19:02:49.770 UTC [91] DEBUG:  write 0/16A3490 flush 0/16A3490 apply 0/16A3490 (reply requested) reply_time 1970-01-01 00:00:00+00
```